common/src/main/java/com/lightstep/tracer/shared/AbstractTracer.java

```diff
@@ -346,19 +346,30 @@ public Boolean flush(long timeoutMillis) {
 
     protected abstract SimpleFuture<Boolean> flushInternal(boolean explicitRequest);
 
+    /**
+     * Do not call this function while holding a lock.
+     *
+     */
     private boolean initializeCollectorClient() {
+        CollectorClient newClient;
         try {
             ManagedChannelBuilder builder =
                 ManagedChannelBuilder.forAddress(collectorURL.getHost(), collectorURL.getPort());
             if (collectorURL.getProtocol() == "http") {
                 builder.usePlaintext(true);
             }
-            client = new CollectorClient(this, builder);
```

Comment:
We call `initializeCollectorClient` lazily during `addSpan` or `sendReportWorker`, I don't think it needs to be synchronized while setting up the connection, just when we write to the `client` variable, with optimistic locking so this warrants another check to ensure `client` is still null before writing to it. 

However, I do see your other point re: setting up client connections a large # of times with this approach, which is wasteful. I'll think about this a bit more and maybe do initialize at setup as you suggested.